fog/sql_recovery_db/src/lib.rs

```diff
@@ -241,6 +242,57 @@ impl RecoveryDb for SqlRecoveryDb {
         Ok(maybe_index.map(|val| val as u64))
     }
 
+    fn get_ingress_key_records(
+        &self,
+        start_block_at_least: u64,
+    ) -> Result<Vec<IngressPublicKeyRecord>, Error> {
+        let conn = self.pool.get()?;
+
+        use schema::ingress_keys::dsl;
+        let key_records: Vec<models::IngressKey> = dsl::ingress_keys
+            .filter(dsl::start_block.ge(start_block_at_least as i64))
+            .load(&conn)?;
+
+        let mut key_records = key_records
+            .iter()
+            .map(
+                |ingress_key_data| -> Result<IngressPublicKeyRecord, Error> {
+                    let status = IngressPublicKeyStatus {
+                        start_block: key_records[0].start_block as u64,
+                        pubkey_expiry: key_records[0].pubkey_expiry as u64,
+                        retired: key_records[0].retired,
+                    };
+
+                    // Look in ingested_blocks table for last scanned block
+                    let ingress_key: CompressedRistrettoPublic =
+                        *ingress_key_data.ingress_public_key;
+                    let key_bytes: &[u8] = ingress_key.as_ref();
+                    use schema::ingested_blocks::dsl;
+                    let maybe_index: Option<i64> = dsl::ingested_blocks
+                        .filter(dsl::ingress_public_key.eq(key_bytes))
+                        .select(diesel::dsl::max(dsl::block_number))
+                        .first(&conn)?;
+
+                    Ok(IngressPublicKeyRecord {
+                        key: ingress_key,
+                        status,
+                        last_scanned_block: maybe_index.map(|x| x as u64),
+                    })
+                },
+            )
+            .collect::<Result<Vec<IngressPublicKeyRecord>, Error>>()?;
+
+        // Sort them as described in get_ingress_key_records trait comments
+        key_records.sort_unstable_by(|a, b| -> Ordering {
```

Comment:
With the subquery in place it might be possible to have the sql server do the ordering.
There's some relevant stuff in https://docs.diesel.rs/diesel/query_dsl/trait.QueryDsl.html#method.order and https://github.com/diesel-rs/diesel/issues/210

I can try doing it as well.
If you want to not spend time on this then at least wrap this whole thing in a read-only transaction (https://docs.diesel.rs/diesel/pg/struct.TransactionBuilder.html#method.read_only)

Reply:
when i started the PR i thought i would need this ordering but in the end nothing relies on it, we could probably punt on it

Reply:
If the ordering isn't needed then I would remove it and the mention in the documentation. That would make the transition from N+1 to 1 query with a subquery more trivial.